ra2ce/network/avg_speed/avg_speed_calculator.py

```diff
@@ -0,0 +1,215 @@
+        """
+        Calculates the average speed from OSM roads, per road type.
+        If the average speed is already calculated and saved in a CSV, it will be read from there.
+        After calculation it will be saved in a CSV file in the output_graph_dir.
+
+        Args: None
+
+        Returns:
+            AvgSpeed: Object with the average speeds per road type
+        """
+        if self.output_graph_dir:
+            _avg_speed_path = self.output_graph_dir.joinpath("avg_speed.csv")
+            if _avg_speed_path.is_file():
+                return AvgSpeedReader().read(_avg_speed_path)
+
+            logging.warning(
+                "No valid file found with average speeds %s, calculating and saving them instead.",
+                _avg_speed_path,
+            )
+
+        _length_array, _maxspeed_array = list(
+            zip(
+                *(
+                    ("length" in e, "maxspeed" in e)
+                    for _, _, e in self.graph.edges.data()
+                )
+            )
+        )
+        if all(_length_array) and any(_maxspeed_array):
+            # Add time weighing - Define and assign average speeds; or take the average speed from an existing CSV
+            _avg_speed = self._get_avg_speed("highway")
+            if self.output_graph_dir:
+                AvgSpeedWriter().export(
+                    self.output_graph_dir.joinpath("avg_speed.csv"), _avg_speed
+                )
+        else:
+            _avg_speed = AvgSpeed()
+            logging.info(
+                "No attributes found in the graph to estimate average speed per network segment."
+            )
+
+        self.avg_speed = _avg_speed
+        return _avg_speed
+
+    def assign(self) -> nx.Graph:
+        """
+        Assigns the average speed and time to roads in an existing (OSM) graph.
+        If not already calculated, the average speed will be calculated first.
+
+        Args: None
+
+        Returns:
+            graph (NetworkX graph): NetworkX graph with an additional attribute 'avgspeed' and 'time'
+        """
+
+        if not self.avg_speed:
+            self.avg_speed = self.calculate()
+
+        # calculate the average maximum speed per edge and assign the ones that don't have a value
+        for u, v, k, edata in self.graph.edges.data(keys=True):
+            _rt = edata.get("highway", None)
+            _length = edata.get("length", None)  # m
+            _speed = self.parse_speed(edata.get("maxspeed", None))
+            if not _length:
+                _length = nut.line_length(edata["geometry"], self.graph.graph["crs"])
+                self.graph.edges[u, v, k]["length"] = _length
+            if not _speed:
+                _speed = self.avg_speed.get_avg_speed(_rt)
+            self.graph.edges[u, v, k]["avgspeed"] = max(
+                round(_speed, 0), 1
+            )  # km/h (at least 1)
+            self.graph.edges[u, v, k]["time"] = round(_length * 1e-3 / _speed, 3)  # h
```

Comment:
Maybe I'm wrong here, but: 
- Avoid comments by the end of a line.
- Simplify code whenever possible
```suggestion
        def get_speed(edata) -> float:
            _rt = edata.get("highway", None)
            _speed = self.parse_speed(edata.get("maxspeed", None))
            if not _speed:
                # We do not define this in the `edata.get(.., fallback_value)` fallback_value
                # as the fallback_value gets always evaluated.
                _speed = self.avg_speed.get_avg_speed(_rt)
            return _speed

        for u, v, k, edata in self.graph.edges.data(keys=True):
            _speed = get_speed(edata)
            _length = edata.get("length", None)  # m
            if not _length:
                # We do not define this in the `edata.get(.., fallback_value)` fallback_value
                # as the fallback_value gets always evaluated.
                _length = nut.line_length(edata["geometry"], self.graph.graph["crs"])
                self.graph.edges[u, v, k]["length"] = _length
            # km/h (at least 1)
            self.graph.edges[u, v, k]["avgspeed"] = max(round(_speed, 0), 1)  
            # h
            self.graph.edges[u, v, k]["time"] = round(_length * 1e-3 / _speed, 3)
```

Reply:
Inline comments are accepted in PEP: https://peps.python.org/pep-0008/#inline-comments

Nice suggestion to extract `get_speed`. I didn't know fallback values get always evaluated.